module/builder/consensus/builder.go

```diff
@@ -442,3 +408,68 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
 
 	return header, nil
 }
+
+// block2SealMap creates a map: blockID -> seal
+// from all seals in our mempool. Its also checks for inconsistent seals:
+// we consider two seals as inconsistent, if they have different start or end states
+func (b *Builder) block2SealMap() (map[flow.Identifier]*flow.IncorporatedResultSeal, error) {
+	// TODO: the following implementation is somewhat temporary and contains a lot of sanity checks
+	//       probably should be cleaned up once we have full sealing
+	encounteredInconsistentSealsForSameBlock := false
+	byBlock := make(map[flow.Identifier]*flow.IncorporatedResultSeal)
+	for _, irSeal := range b.sealPool.All() {
+		irSealInitialState, ok := irSeal.IncorporatedResult.Result.InitialStateCommit()
+		if !ok || len(irSealInitialState) < 1 { // missing or empty initial state commitment
+			// fatal error: respective Execution Result should have been rejected by matching engine
+			return nil, fmt.Errorf("seal for result without initial state: %v", irSeal.IncorporatedResult.Result.ID())
+		}
+		irSealFinalState, ok := irSeal.IncorporatedResult.Result.FinalStateCommitment()
+		if !ok || len(irSealFinalState) < 1 { // missing or empty final state commitment
+			// fatal error: respective Execution Result should have been rejected by matching engine
+			return nil, fmt.Errorf("seal for result without final state: %v", irSeal.IncorporatedResult.Result.ID())
+		}
+		if !bytes.Equal(irSeal.Seal.FinalState, irSealFinalState) {
+			// fatal error: matching engine has constructed seal with inconsistent values
+			return nil, fmt.Errorf("inconsistent final sate in result and seal for block %v", irSeal.Seal.BlockID)
+		}
```

Comment:
IMO, I think these sanity check could just be moved to matching engine. 

Otherwise, we are either missing checks in matching engine or duplicating checks here.